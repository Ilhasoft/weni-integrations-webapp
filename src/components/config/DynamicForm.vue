<template>
  <div class="dynamic-form">
    <div class="dynamic-form__fields" v-for="(input, index) in inputs" :key="index">
      <unnnic-input
        v-if="input.type === 'input'"
        ref="unnnic-input"
        :class="[!input.label && 'dynamic-form__fields--top-margin', 'dynamic-form__fields__input']"
        :type="getType(input)"
        v-model="input.value"
        :label="input.label && getValue(input, 'label')"
        :placeholder="input.placeholder && getValue(input, 'placeholder')"
        :message="input.message && getValue(input, 'message')"
        @update:modelValue="(e) => emitInput(index, e)"
      />

      <div v-if="input.type === 'select'">
        <unnnic-label :label="input.label && getValue(input, 'label')" />
        <unnnic-select-smart ref="unnnic-select" :options="input.options" v-model="input.value" />
      </div>
      <div v-else-if="input.type === 'upload'">
        <unnnic-label :label="input.label && getValue(input, 'label')" />
        <unnnic-upload-area
          ref="unnnic-upload"
          :files="input.props.files"
          :acceptMultiple="input.props.acceptMultiple"
          :supportedFormats="input.props.supportedFormats"
          :maximumUploads="input.props.maximumUploads"
          :maxFileSize="input.props.maxFileSize"
          :filesProgress="input.props.filesProgress"
          :isUploading="input.props.isUploading"
          :canImport="input.props.canImport"
          :canDelete="input.props.canDelete"
          :shouldReplace="input.props.shouldReplace"
          @fileChange="(e) => emitInput(index, e)"
        />
      </div>
      <unnnic-checkbox
        ref="unnnic-checkbox"
        v-else-if="input.type === 'checkbox'"
        class="dynamic-form__fields--top-margin"
        v-model="input.value"
        :textRight="input.label && getValue(input, 'label')"
        @change="(e) => emitInput(index, e)"
      />
    </div>
  </div>
</template>

<script>
  export default {
    name: 'DynamicForm',
    props: {
      inputs: {
        type: Array,
        default: /* istanbul ignore next */ () => [],
      },
      channelCode: {
        type: String,
        default: '',
      },
    },
    methods: {
      emitInput(index, event) {
        const type = this.inputs[index]?.type;
        switch (type) {
          case 'select':
            this.$emit('input', { index, value: event[0] });
            break;
          default:
            this.$emit('input', { index, value: event });
            break;
        }
      },
      getType(input) {
        return input.error ? 'error' : 'normal';
      },
      getValue(input, type) {
        const key = `channels.inputs.${this.channelCode}.${input.name}.${type}`;
        return this.$te(key) ? this.$t(key) : this.$t(input[type]);
      },
    },
  };
</script>

<style lang="scss" scoped>
  .dynamic-form {
    &__fields {
      &--top-margin {
        margin-top: $unnnic-spacing-stack-xs;
      }

      &__input {
        ::v-deep {
          .unnnic-form__message {
            color: $unnnic-color-neutral-cloudy;
          }
        }
      }
    }
  }
</style>
