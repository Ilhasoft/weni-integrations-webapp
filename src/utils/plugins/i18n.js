import Vue from 'vue';
import VueI18n from 'vue-i18n';

Vue.use(VueI18n);

import en from '../../locales/en.json';
import pt_br from '../../locales/pt_br.json';
import es_es from '../../locales/es_es.json';

const languages = {
  'en-us': en,
  'pt-br': pt_br,
  es: es_es,
};

const messages = Object.assign(languages);

const i18n = new VueI18n({
  locale: 'en-us',
  fallbackLocale: 'en-us',
  messages,
});

export default i18n;
