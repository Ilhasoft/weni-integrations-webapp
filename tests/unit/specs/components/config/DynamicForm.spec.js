import { shallowMount, createLocalVue } from '@vue/test-utils';
import DynamicForm from '@/components/config/DynamicForm.vue';
import i18n from '@/utils/plugins/i18n';

const localVue = createLocalVue();

describe('DynamicForm.vue', () => {
  let wrapper;
  beforeEach(() => {
    const inputs = [
      {
        type: 'input',
        name: 'input',
        label: 'input',
        value: null,
      },
      {
        type: 'select',
        name: 'select',
        label: 'select',
        placeholder: 'select',
        value: null,
        options: [
          { value: 'select1', text: 'select1' },
          { value: 'select2', text: 'select2' },
        ],
      },
    ];

    wrapper = shallowMount(DynamicForm, {
      localVue,
      i18n,
      propsData: {
        inputs,
      },
      mocks: {
        $t: () => 'some specific text',
      },
      stubs: {
        UnnnicInput: true,
        UnnnicSelect: true,
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should be rendered properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call emitInput on input component event', async () => {
    const spy = spyOn(wrapper.vm, 'emitInput');
    const input = wrapper.findComponent({ ref: 'unnnic-input' });

    expect(spy).not.toHaveBeenCalled();
    input.vm.$emit('input', 'test');
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith(0, wrapper.props().inputs[0], 'test');
  });

  it('should call emitInput on select component event', async () => {
    const spy = spyOn(wrapper.vm, 'emitInput');
    const input = wrapper.findComponent({ ref: 'unnnic-select' });

    expect(spy).not.toHaveBeenCalled();
    input.vm.$emit('input', 2);
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith(1, wrapper.props().inputs[1], 2);
  });

  describe('emitInput()', () => {
    it('should emit input value if default type', () => {
      const index = 0;
      const input = wrapper.props().inputs[index];
      const event = 'test';
      expect(wrapper.emitted('input')).toBeFalsy();
      wrapper.vm.emitInput(index, input, event);
      expect(wrapper.emitted('input')).toBeTruthy();
      expect(wrapper.emitted('input')[0]).toEqual([{ index, value: event }]);
    });
    it('should emit option value if input type is select', () => {
      const index = 1;
      const input = wrapper.props().inputs[index];
      const event = 'select2';
      expect(wrapper.emitted('input')).toBeFalsy();
      wrapper.vm.emitInput(index, input, event);
      expect(wrapper.emitted('input')).toBeTruthy();
      expect(wrapper.emitted('input')[0]).toEqual([{ index, value: input.options[1].value }]);
    });
  });
});
