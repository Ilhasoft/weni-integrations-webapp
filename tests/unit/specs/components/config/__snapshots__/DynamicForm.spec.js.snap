// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicForm.vue should be rendered properly 1`] = `
<div class="dynamic-form">
  <div class="dynamic-form__fields">
    <unnnicinput-stub type="normal" label="input" class="dynamic-form__fields__input"></unnnicinput-stub>
  </div>
  <div class="dynamic-form__fields">
    <unnnicselect-stub placeholder="select" label="select">
      <option value="select1">
        select1
      </option>
      <option value="select2">
        select2
      </option>
    </unnnicselect-stub>
  </div>
</div>
`;
